pkg/ring/ring.go

```diff
@@ -515,6 +512,12 @@ func (r *Ring) ShuffleShard(identifier string, size int) ReadRing {
 			tokens = r.ringTokens
 		}
 
+		// Initialise the random generator used to select instances in the ring.
+		// Since we consider each zone like an independent ring, we have to use dedicated
+		// pseudo-random generator for each zone, in order to guarantee the "consistency"
+		// property when the shard size changes or a new zone is added.
+		random := rand.New(rand.NewSource(util.ShuffleShardSeed(identifier + zone)))
```

Comment:
You can save allocations by passing both strings as individual arguments, and modifying `ShuffleShardSeed` to either take two arguments instead.

Reply:
Also `identifier + zone` will produce same string for different tenants:

* tenant "A" zone "BC"
* tenant "AB", zone "C"

This is solved by adding separator constant between values.

Reply:
> You can save allocations by passing both strings as individual arguments, and modifying ShuffleShardSeed to either take two arguments instead.

I tried but apparently doesn't reduce allocations because of the `[]byte` slice we have to allocate anyway. Thoughts?

> Also identifier + zone will produce same string for different tenants:

True. I'm just wondering if it's realistic that zones will have different format. If I think about AWS or GCP, zones within the same region have the same pattern (eg. `us-central1-a`, `us-central1-b`, `us-central1-c`, ...).

Reply:
> I tried but apparently doesn't reduce allocations because of the []byte slice we have to allocate anyway. Thoughts?

Yes, that's annoying. We could use `yoloBuf` as we know that buffers are not retained in md5 digest code. Would make benchmarks look good :-)

> True. I'm just wondering if it's realistic that zones will have different format. If I think about AWS or GCP, zones within the same region have the same pattern (eg. `us-central1-a`, `us-central1-b`, `us-central1-c`, ...).

Not a big concern, but it's low effort, esp. if we send both values as arguments, we can easily write constant separator between them to hasher.

Reply:
I used `YoloBuf`.

Memory allocations diff **without** `YoloBuf`:

```
benchmark                                                                              old allocs     new allocs     delta
BenchmarkRing_ShuffleShard/num_instances_=_50,_num_zones_=_1,_shard_size_=_3-12        25             25             +0.00%
BenchmarkRing_ShuffleShard/num_instances_=_50,_num_zones_=_1,_shard_size_=_10-12       27             27             +0.00%
BenchmarkRing_ShuffleShard/num_instances_=_50,_num_zones_=_1,_shard_size_=_30-12       31             31             +0.00%
BenchmarkRing_ShuffleShard/num_instances_=_50,_num_zones_=_3,_shard_size_=_3-12        43             51             +18.60%
BenchmarkRing_ShuffleShard/num_instances_=_50,_num_zones_=_3,_shard_size_=_10-12       49             57             +16.33%
BenchmarkRing_ShuffleShard/num_instances_=_50,_num_zones_=_3,_shard_size_=_30-12       55             63             +14.55%
BenchmarkRing_ShuffleShard/num_instances_=_100,_num_zones_=_1,_shard_size_=_3-12       25             25             +0.00%
BenchmarkRing_ShuffleShard/num_instances_=_100,_num_zones_=_1,_shard_size_=_10-12      27             27             +0.00%
BenchmarkRing_ShuffleShard/num_instances_=_100,_num_zones_=_1,_shard_size_=_30-12      31             31             +0.00%
BenchmarkRing_ShuffleShard/num_instances_=_100,_num_zones_=_3,_shard_size_=_3-12       43             51             +18.60%
BenchmarkRing_ShuffleShard/num_instances_=_100,_num_zones_=_3,_shard_size_=_10-12      49             57             +16.33%
BenchmarkRing_ShuffleShard/num_instances_=_100,_num_zones_=_3,_shard_size_=_30-12      55             63             +14.55%
BenchmarkRing_ShuffleShard/num_instances_=_1000,_num_zones_=_1,_shard_size_=_3-12      25             25             +0.00%
BenchmarkRing_ShuffleShard/num_instances_=_1000,_num_zones_=_1,_shard_size_=_10-12     27             27             +0.00%
BenchmarkRing_ShuffleShard/num_instances_=_1000,_num_zones_=_1,_shard_size_=_30-12     31             31             +0.00%
BenchmarkRing_ShuffleShard/num_instances_=_1000,_num_zones_=_3,_shard_size_=_3-12      43             51             +18.60%
BenchmarkRing_ShuffleShard/num_instances_=_1000,_num_zones_=_3,_shard_size_=_10-12     49             57             +16.33%
BenchmarkRing_ShuffleShard/num_instances_=_1000,_num_zones_=_3,_shard_size_=_30-12     55             63             +14.55%
```

Memory allocations diff **with** `YoloBuf`:

```
benchmark                                                                              old allocs     new allocs     delta
BenchmarkRing_ShuffleShard/num_instances_=_50,_num_zones_=_1,_shard_size_=_3-12        25             24             -4.00%
BenchmarkRing_ShuffleShard/num_instances_=_50,_num_zones_=_1,_shard_size_=_10-12       27             26             -3.70%
BenchmarkRing_ShuffleShard/num_instances_=_50,_num_zones_=_1,_shard_size_=_30-12       31             30             -3.23%
BenchmarkRing_ShuffleShard/num_instances_=_50,_num_zones_=_3,_shard_size_=_3-12        43             48             +11.63%
BenchmarkRing_ShuffleShard/num_instances_=_50,_num_zones_=_3,_shard_size_=_10-12       49             54             +10.20%
BenchmarkRing_ShuffleShard/num_instances_=_50,_num_zones_=_3,_shard_size_=_30-12       55             60             +9.09%
BenchmarkRing_ShuffleShard/num_instances_=_100,_num_zones_=_1,_shard_size_=_3-12       25             24             -4.00%
BenchmarkRing_ShuffleShard/num_instances_=_100,_num_zones_=_1,_shard_size_=_10-12      27             26             -3.70%
BenchmarkRing_ShuffleShard/num_instances_=_100,_num_zones_=_1,_shard_size_=_30-12      31             30             -3.23%
BenchmarkRing_ShuffleShard/num_instances_=_100,_num_zones_=_3,_shard_size_=_3-12       43             48             +11.63%
BenchmarkRing_ShuffleShard/num_instances_=_100,_num_zones_=_3,_shard_size_=_10-12      49             54             +10.20%
BenchmarkRing_ShuffleShard/num_instances_=_100,_num_zones_=_3,_shard_size_=_30-12      55             60             +9.09%
BenchmarkRing_ShuffleShard/num_instances_=_1000,_num_zones_=_1,_shard_size_=_3-12      25             24             -4.00%
BenchmarkRing_ShuffleShard/num_instances_=_1000,_num_zones_=_1,_shard_size_=_10-12     27             26             -3.70%
BenchmarkRing_ShuffleShard/num_instances_=_1000,_num_zones_=_1,_shard_size_=_30-12     31             30             -3.23%
BenchmarkRing_ShuffleShard/num_instances_=_1000,_num_zones_=_3,_shard_size_=_3-12      43             48             +11.63%
BenchmarkRing_ShuffleShard/num_instances_=_1000,_num_zones_=_3,_shard_size_=_10-12     49             54             +10.20%
BenchmarkRing_ShuffleShard/num_instances_=_1000,_num_zones_=_3,_shard_size_=_30-12     55             60             +9.09%
```